main/src/library/MutHeap.flix

```diff
@@ -0,0 +1,240 @@
+        StringBuilder.toString(sb)
+    }
+
+    ///
+    /// Returns an iterator over MutHeap `h`
+    ///
+    /// Modifying `h` during iteration is undefined
+    ///
+    pub def iterator(rc: Region[r1], h: MutHeap[a, r2]): Iterator[a, r1 + r2, r1] \ {r1, r2} =
+        let MutHeap(_, arr, s) = h;
+        let numElems = deref s;
+        let idx = ref 0 @ rc;
+        let next = () -> {
+            let i = deref idx;
+            if (i < numElems) {
+                let elem = Array.get(i, deref arr);
+                Some(elem)
+            }
+            else {
+                None
+            }
+        };
+        Iterator.iterate(rc, next)
+
+    ///
+    /// Reinforces the max heap invariant after an element is added
+    ///
+    def heapifyUp(idx: Int32, h: MutHeap[a, r]): Unit \ r with Order[a] =
+        let MutHeap(_, arr, _) = h;
+        if (idx != 0) {
+            let parentIdx = (idx - 1) / 2;
+            let cur  = Array.get(idx, deref arr);
+            let parent = Array.get(parentIdx, deref arr);
+            if (cur > parent) {
+                Array.put(parent, idx, deref arr);
+                Array.put(cur, parentIdx, deref arr);
+                heapifyUp(parentIdx, h)
+            }
+            else ()
+        }
+        else ()
+
+    ///
+    /// Reinforces the max heap invariant after an element is removed
+    ///
+    def heapifyDown(idx: Int32, h: MutHeap[a, r]): Unit \ r with Order[a] =
+        let MutHeap(_, arr, s) = h;
+        let size = deref s;
+        let lChildIdx = idx * 2 + 1;
+        let rChildIdx = idx * 2 + 2;
+        let cur = Array.get(idx, deref arr);
+        if (size >= rChildIdx) {
+            if (size == rChildIdx) {
+                let child = Array.get(lChildIdx, deref arr);
+                if (Order.less(cur, child)) {
+                    Array.put(child, idx, deref arr);
+                    Array.put(cur, lChildIdx, deref arr)
+                }
+                else ()
+            }
+            else {
+                let lChild = Array.get(lChildIdx, deref arr);
+                let rChild = Array.get(rChildIdx, deref arr);
+                if ((lChild > cur) or (rChild > cur)) {
+                    if (lChild > rChild) {
+                        Array.put(cur, lChildIdx, deref arr);
+                        Array.put(lChild, idx, deref arr);
+                        heapifyDown(lChildIdx, h)
+                    }
+                    else {
+                        Array.put(cur, rChildIdx, deref arr);
+                        Array.put(rChild, idx, deref arr);
+                        heapifyDown(rChildIdx, h)
+                    }
+                }
+                else ()
+            }
+        }
+        else ()
+
+    ///
+    /// Expands the internal array of MutHeap `h` if the array capacity is being approached
+    ///
+    /// Expansion transpires only when the number of elements in `h` is 1 less than the array capacity
```

Comment:
Does this mean we never use the last array slot? Why not when it is equal to the array capacity?

Reply:
Good point